### PR TITLE
AUDIO: Fix playback in OPL3 mono mode

### PR DESCRIPTION
**Pull request by Raffaello**
If running an OPL2 track on a DOS_BOX OPL3 emulator, The buffer will be divided by 2 like if it is stereo, but it will generate a mono audio stream, as opl3Active flag is disabled, so it will act as an OPL2.
The error results in wrong sound produced.
This fix the issue when OPL3 is used as an OPL2 and therefore as a mono audio source.

The current error should be easily reproducible if using OPL3 with .ADL Westwood's files.

This bug it might have never been noticed because those files, in kyra engine, are using always OPL2 (kAdlib) is used.

### CODE REVIEW

**Comment by Raffaello** [comment]
actullally it should be something like this:
(i am not updating the PR as it might just never been merged as directly on ScummVM is not noticeable the small error and also further improvement are in development and i might not be aware of)

```
if(isStereo())
{
    while (length_ > 0)
    {
        const unsigned int readSamples = std::min<unsigned int>(length_, bufferLength);
        const unsigned int readSamples2 = (readSamples << 1);
        if (m_emulator->opl3Active)
        {
            m_emulator->GenerateBlock3(readSamples, tempBuffer.data());
            for (unsigned int i = 0; i < readSamples2; ++i)
                buffer[i] = static_cast<int16_t>(tempBuffer[i]);
        }
        else
        {
            m_emulator->GenerateBlock2(readSamples, tempBuffer.data());
            for (unsigned int i = 0, j =0; i < readSamples; ++i, j+=2)
            {
                buffer[j] = buffer[j+1] = static_cast<int16_t>(tempBuffer[i]);
            }
        }

        buffer += static_cast<int16_t>(readSamples2);
        length_ -= readSamples;
    }
}
else {
....
}
```

the main branch if checks if it is `isStereo()` this is required due to the mixer is initialized as stereo if not an OPL2.

within that branch if OPL3 is working as OPL2, must call the `GenerateBlock2` for OPL2 compatiblity and interleave the Left and Right channels generated by the mono OPL2 to keep it sounding right.
This is required i think due to some DosBox internal channel masks as well if going looking deeper within the generating sounds.

It looks silly and overcompiclated but an OPL3 in this way can be used to run OPL2 sounds (otherwise must be initialized a DosBox OPL2 with OPL2 drivers)

anyway.
I thought it might be eventually interesting, but in ScummVM as the right OPL version is initialized for each game this small patch won't be required at the very end. 
And if not using DosBox OPL emulator, for instance if using Nuked, there is no issue internally as it is fully opl2 compatible, only dosbox is kind of segmented among OPL2 and OPL3.

**Comment by antoniou79** [comment]
Hello, and my apologies. I think I've sent something related to another
project in this e-mail thread by mistake.

I'm sorry for any confusion I may have caused.

All the best,
Th.

**Comment by sev-** [comment]
@Raffaello okay, it makes sense. Could you however change the git log message to 

```
AUDIO: Fix playback in OPL3 mono mode

.... here goes your extensive comment as is
```

**Comment by Raffaello** [comment]
> @Raffaello okay, it makes sense. Could you however change the git log message to
> 
> ```
> AUDIO: Fix playback in OPL3 mono mode
> 
> .... here goes your extensive comment as is
> ```
@sev-  It makes sense the 1 line code change or the snippet of the code in the comment that i also posted? 

This 1 line change, **if i recall correctly**, it works ok in OPL2 and OPL3, but it does mess up DUAL_OPL mode

hence the latter longer patch posted in the comment.

**Comment by sev-** [comment]
@Raffaello  Okay, clarifying. According to our code commit guidelines, the format is:

```
SUBSTSEM: One-line description

Detailed description
It could be multiline.
```

The subsystem in your case is `AUDIO`. I think you know better how to describe it in one line.

**Comment by Raffaello** [comment]
@sev-  sorry for the misunderstanding. I understood your comment. I did what you asked for.

Besides I was asking and writing down about a clarification on the actual implemented patch and this PR: 
I posted a comment with a fragment of code that should be a better fix, but i didn't implemented because i think doesn't really add too much value in ScummVM, because for each game the OPL emulators are intialized as expected.

Anyway i am reporting here that fragment of code for an eventually further code review
(it is not a code to be fit as it is in ScummVM as it doesn't follow the C++ style guide, so take it more from its logic perspective):

```cpp
if(isStereo()) // _type != Config::kOpl2
{
    while (length_ > 0)
    {
        const unsigned int readSamples = std::min<unsigned int>(length_, bufferLength);
        const unsigned int readSamples2 = (readSamples << 1);
        if (m_emulator->opl3Active) // basically if an OPL3 in OPL3 mode [1]
        {
            m_emulator->GenerateBlock3(readSamples, tempBuffer.data());
            for (unsigned int i = 0; i < readSamples2; ++i)
                buffer[i] = static_cast<int16_t>(tempBuffer[i]);
        }
        else  // like the case OPL3 running as OPL2 [2]
        {
            m_emulator->GenerateBlock2(readSamples, tempBuffer.data());
            for (unsigned int i = 0, j =0; i < readSamples; ++i, j+=2)
            {
                buffer[j] = buffer[j+1] = static_cast<int16_t>(tempBuffer[i]);
            }
        }

        buffer += static_cast<int16_t>(readSamples2);
        length_ -= readSamples;
    }
}
else {
....
}
```

This fragment should cover all the scenarios, with OPL3 running OPL2 mode, and cover also DUAL_OPL2. 
At least i had to further evolving this patch because i had problem with DUAL_OPL2 or OPL3 with Westwood ADLDriver when using DosBox OPL Emulation (ADLDriver is a OPL2 Driver basically).

Basically this current PR 1 line code change might have issues with DUAL_OPL2, but again in ScummVM might not be really noticeable, because using ADLDriver (Westwood game [kyrandia, etc....] ) when using those games is by default using a OPL2 (AdLib) emulation. Instead i am doing some experimantation to use that Driver with a OPL3 or DUAL_OPL2 and that were this unnoticiable "bug" came to surface.
NOTE related to ADLDriver:
- [1] In the case using DosBox DUAL_OPL is going on the isStereo() and opl3Active path
- [2] in the case using DosBox OPL3 is going on the isStereo() and NOT opl3Active path (else branch of the internal if)

**Comment by sev-** [comment]
Okay, please go with a more comprehensive solution, @Raffaello 

It is better to avoid possible side-effects.

**Comment by sev-** [comment]
@Raffaello Thank you for updating the code. However, we require all the contributions to follow our [Code Formatting conventions](https://wiki.scummvm.org/index.php?title=Code_Formatting_Conventions). Please, fix.

**Comment by Raffaello** [comment]
@sev- i think i have done the code style correctly. if it wouldn't be the case please let me know. thanks

**Comment by lephilousophe** [comment]
Due to a mistake of mine, the PR got closed.
This has been merged manually in 15425aa66be
